### ISSUE TITLE
assembly is overzealous in reporting some kinds of type error.

### ISSUE THREAD

**Issue by jannotti** [algorand/go-algorand#3827]
```
func TestTypeComplaints(t *testing.T) {
	testProg(t, "int 1; return; store 0", AssemblerMaxVersion)
}
```
fails because the type stack seems wrong for that `store` but it's unreachable.

**Comment by jannotti** [comment]
Giving more detail - the assembler today tracks type errors, but only reports an _error_ if the mismatch occurs before before any branches in the program.  The assembler can't properly analyze whether a mismatch in stack expectations is real if branches have occurred.  This program has no apparent branches, so it reports this as an error.  However, it is probably useful to consider `return` to be a branch.  This program runs and reports approval just fine,  It should be ok to have extraneous instructions at the end of a program that can never be reached (some have proposed encoding program metadata this way).

There is an old PR #2762, that tries to report errors in more detail, by noticing type mismatches that are sure to occur because they depend only on the instructions within a single basic block.  It would be great to have that check happening, but even that is done, that kind of check should not be an error if it occurs in dead code.

It's always worth saying that detecting dead code as an error might be a useful thing. In at least one real-world case, warning about dead code would have prevented an exploit (that was fortunately fixed before exploited). A nice solution that errors on unintentional dead code, but allows intentional dead code (the aforementioned metadata) is way beyond the scope of this issue.

**Comment by jannotti** [comment]
Fixed by #3870
